service/rest.go

```diff
@@ -0,0 +1,99 @@
+package service
+
+import (
+	"net/http"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/auth"
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/gimlet"
+	"github.com/pkg/errors"
+	"github.com/urfave/negroni"
+)
+
+// restContextKey is the type used to store
+type restContextKey int
+
+const RestContext restContextKey = 0
+
+type restAPIService interface {
+	GetSettings() evergreen.Settings
+	LoggedError(http.ResponseWriter, *http.Request, int, error)
+}
+
+type restAPI struct {
+	restAPIService
+}
+
+type restV1middleware struct {
+	restAPIService
+}
+
+func (ra *restV1middleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	vars := gimlet.GetVars(r)
+	ctx, err := model.LoadContext(vars["task_id"], vars["build_id"], vars["version_id"], vars["patch_id"], vars["project_id"])
+	if err != nil {
+		// Some database lookup failed when fetching the data - log it
+		ra.LoggedError(rw, r, http.StatusInternalServerError, errors.Wrap(err, "Error loading project context"))
+		return
+	}
+
+	if ctx.ProjectRef != nil && ctx.ProjectRef.Private && GetUser(r) == nil {
```

Comment:
this feels like it belongs elsewhere, either in an auth method or in the route itself

Reply:
I think better to keep this out of the route itself, and also, I totally agree with you, but the problem is that this has to happen really late in the process (on the whole), so it might not get much better than this.